### ISSUE TITLE
Making writing easier and more intuitive

### ISSUE THREAD

**Issue by amywtlin** [forem/forem#14807]

**Comment by aitchiss** [comment]
>> Initially, we viewed the toolbar as a stepping stone towards WYSIWYG but I don't believe we are going to be able to re-use any aspect of it other than the icons/layout which is a pretty minimal part of the work.

>Can you elaborate on that? Again, I was under the impression that this is actually what we CAN do and that's literally why we decided to keep editor a separate/standalone project :).

Yes for sure. Since first discussions (and as far as I know, _how_ we are doing WYSIWYG is still an open question/discussion), it's becoming more clear to me that it's really only the UI of the toolbar that we are going to be able to reuse. I'm also going to preface this by saying I haven't had a great deal of time to research any of this super thoroughly... which is kind of one of the points I'm making above.

The top-level issue is that I don't believe a string of markdown alone is sufficient to power a WYSIWYG experience. There are plenty of packages that will convert markdown to HTML for us but realistically we don't want to be calling a parser like this on every key stroke or toolbar button press. There would also just be a huge amount of complexity in making those toolbar buttons work in the expected way - e.g. on a very basic level, how do we work out cursor position vs position in the markdown string when markdown syntax can only be interpreted by additional characters in the code. I think we would be setting ourselves an impossible task.

If we want to continue saving posts as markdown, I believe we are going to need a schema _in between_ the markdown string and the WYSIWYG experience. Even if we use a 3rd party library for nothing else, we should be looking to use one for this purpose. [Prosemirror](https://prosemirror.net/examples/markdown/) would be a good starting point for investigation (we briefly discussed this in a slack thread). With this 'in between' schema we will want a different set of actions to be executed on toolbar button press (e.g. amend the given node in that schema with some kind of formatting). Working with that schema is going to be much easier to do things like 'undo this formatting on button press' and 'continue a bulleted list as you press enter'.

My understanding was also that we would still _explore_ 3rd party libraries for the full experience, rather than completely write them off from day one (even if we are starting out sceptical that they will meet our needs). I've said it before, but the size of this task if we don't use a 3rd party lib is huge.

**Comment by ludwiczakpawel** [comment]
> Yes for sure. Since first discussions (and as far as I know, how we are doing WYSIWYG is still an open question/discussion), it's becoming more clear to me that it's really only the UI of the toolbar that we are going to be able to reuse.

With this point I must disagree :D I think having this toolbar as it is right now in your PR will be huge addition to our current markdown editor. I'd personally ship it and just let it live. 

But with that in mind, I'm not an engineer so I will obviously rely on your (and others) expertise re how to proceed. I'd love everyone to chime in here, and if necessary do some tech research / engineering evaluation so we have better idea on how to tackle Composer Updates and what the best path is (should we continue work on the toolbar for current editor? or should we ditch it and jump straight to WYSIWYG project?). cc @amywtlin @rhymes @citizen428 @msarit

Two big questions are (and these are related obviously so there may be one answer for both):
1. Do we ship Suzanne's toolbar to enhance current editor experience?
2. How should we tackle WYSIWYG?

**Comment by amywtlin** [comment]
> How should we tackle WYSIWYG?

After quite a bit of discussion in the product team, all things considered, we think **it's best that we put WYSIWYG editor on hold**, until we have further information that suggests we should resume work on this. That means, within this epic, we would only finish adding a toolbar and simplifying liquid tags.

> Suzanne rightly pointed out that there are technical challenges to both approaches and we need further research in some aspects but it's also clearer and clearer to me that it really depends on what's the end goal.

@rhymes brought up a really good point here. We set out to **make our editor more intuitive**, as the original title of this epic.  Considering all the extra considerations & maintenances building a WYSIWYG would bring, the amount it would contribute to "making our editor more intuitive" is too marginal to justify the effort.

Special thanks to @aitchiss for raising the issue and to everyone who weighed in. 

-------
> Do we ship Suzanne's toolbar to enhance current editor experience?

It would depend on @aitchiss's assessment here. My thought would be if it takes a few more days to complete the undo experience, then I am fine waiting a few more days for a more complete experience. If it takes weeks, then I'll say let's ship first and patch later.

**Comment by aitchiss** [comment]
> Btw. I'd rather prioritize adding image uploader to this toolbar rather than undo functionality.

TBH I have the majority of image uploader working locally 🙂  I'm not going to add it while we're still only in "storybook world", just because the feature can't be fully tested there (since it's not just pure frontend functionality).

If we want 'undo' functionality before we launch it into the editor I'd suggest we:

- get the current PR into storybook
- get undo functionality into storybook
- launch in editor with the image upload functionality
- review other suggestions that have come up in the course of reviews (e.g. not splitting words when we insert a syntax) and any user testing

If that sounds reasonable I'll add the first 3 things as tasks to the [toolbar issue](https://github.com/forem/forem/issues/14808), and start a running list of other potential improvements there too

**Comment by aitchiss** [comment]
Personally, I don't think so. I think these items listed above are "nice to haves" / "potential future improvements". But I'm happy to defer to what you and @ludwiczakpawel think from a product point of view. I agree though that it makes sense to add the "embed" to the toolbar first (which should be a quick job).

In any case, if we're committed to doing these things _at some point_,  should we turn those tasks into issues? If I'm honest, I'd totally lost sight of these ideas 😅

**Comment by amywtlin** [comment]
Yes. My intention is to have a list of tasks we agree upon, and turn them into issues so we can keep track of them!

**Comment by amywtlin** [comment]
Re: what to do before we bring toolbar to the comment area, FYI I have created 2 new issues (#16780 & #16781 ) and updated the main body of this issue.

**Comment by ludwiczakpawel** [comment]
Potentially useful: 
- https://github.com/forem/forem-internal-eng/issues/354

**Comment by amywtlin** [comment]
@aitchiss I am revisiting this epic issue and evaluating if I can close this. While confirming with Pawel if we are revisiting the idea of wysiwyg editor (because of [this thread](https://forem-team.slack.com/archives/C7GE84DEJ/p1652361427648529)), I think the last thing to do is to turn the rest of items without an issue into issues, and then we can close this epic.

I am wondering if you can help me turn these into issues so we can open them up for contribution as you probably know the details of each of them. 

(I know, I know, this is so long ago 😂 )

**Comment by aitchiss** [comment]
Hey @amywtlin! Yes I can turn the remaining tasks into issues 🙂  Some of them might need fleshed out a bit more before they're ready for development but I'll see where I get to and let you know!

**Comment by amywtlin** [comment]
Thanks so much! Yes, please ping me any time if you need me and let's flesh out things ❤️

**Comment by aitchiss** [comment]
I've converted those tasks to issues now 😄  NB a couple I think need product input (e.g. deciding what shortcuts should be), but I've add labels to them. I also initially added the "Forem team" label but then removed that because I reckon they could be nice community contributions too

**Comment by mikeydorje** [comment]
Closing this epic as completed/outdated as a tracker.
Most scoped items have shipped (toolbar in editor/comments, embed button, undo/shortcuts fixes, mobile overflow work), and remaining items were intentionally split into standalone issues (with several later closed as not planned).
